File: queue/QueueImpByLinkList.py
'''
Created on 2018年8月4日
https://www.geeksforgeeks.org/queue-set-2-linked-list-implementation/
@author: Viking
'''
# python3 program to demonstrate linked list
# based implementation of queue


# A linked list(LL) node
# to store a queue entry
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


# The queue,front stores the front node
# of LL and rear stores the last node of LL
class Queue:
    def __init__(self):
        self.rear = self.front = None

    def isEmpty(self):
        return self.front == None

    def enqueue(self, item):
        tmp = Node(item)

        if self.rear == None:
            self.rear = self.front = tmp
            return

        self.rear.next = tmp
        self.rear = tmp

    def dequeue(self):
        if self.isEmpty():
            return

        tmp = self.front
        self.front = self.front.next

        if self.front == None:
            self.rear = None

        return str(tmp.data)


# Driver code
if __name__ == '__main__':
    queue = Queue()
    queue.enqueue(10)
    queue.enqueue(20)
    queue.dequeue()
    queue.dequeue()
    queue.enqueue(30)
    queue.enqueue(40)
    queue.enqueue(50)
    print("Dequeued item is " + queue.dequeue())


'''
Dequeued item is 30
'''

File: queue/QueueImpByLinkList.java
package org.q10viking.queue;

//A linked list node to store a queue entry
//链表的核心是Node
class QNode{
	int key;
	QNode next;
	QNode(int key){
		this.key = key;
		this.next = null;
	}
	
}

//The queue,front stores the front node of LL and rear stores the 
//Last node of LL(Link list)
public class QueueImpByLinkList {

	QNode rear,front;
	public QueueImpByLinkList() {
		this.front = this.rear = null;
	}
	
	//Method to add an key to the queue
	public void enqueue(int key) {
		//create a new LL node
		QNode tmp = new QNode(key);
		//If queue is empty,then new node is front and rear both
		if(this.rear==null) {
			this.front=this.rear=tmp;
			return ;
		}
		//add the new node at the end of queue and change rear
		this.rear.next = tmp;
		this.rear = tmp;
		
	}
	//Method to remove an key from queue
	public int dequeue() {
		if(this.front == null)
			return 0;
		QNode tmp = this.front;
		this.front = this.front.next;

		//if front becomes null,then change rear also null
		if(this.front==null) {
			this.rear = null;
		}
		return tmp.key;
	}
	
	
	public static void main(String[] args) {
		QueueImpByLinkList queue = new QueueImpByLinkList();
		queue.enqueue(10);
		queue.enqueue(20);
		queue.enqueue(30);
		System.out.println(queue.dequeue());
		System.out.println(queue.dequeue());
		System.out.println(queue.dequeue());
	}
}

/*
10
20
30

*/